ui/src/ethereum/walletConnect.ts

```diff
@@ -0,0 +1,179 @@
+  getConnection(): Connection | undefined;
+
+  // Start the connection flow by showing the modal with the connection
+  // data. Returns `true` if the connection has been succesfully
+  // established and `false` if the user closed the modal without
+  // connecting.
+  connect(): Promise<boolean>;
+
+  disconnect(): Promise<void>;
+
+  // Sign a message with the key for `address`. `message` must be
+  // hex-encoded.
+  signMessage(address: string, message: string): Promise<string>;
+  // Submit a transaction to the network via the wallet and return the
+  // transaction hash. Requires user confirmation. Throws when no
+  // wallet is connected.
+  sendTransaction(tx: ITxData): Promise<string>;
+  // Have the user sign a transaction and return the signed transaction
+  // encoded as hex string. Throws when no wallet is connected.
+  signTransaction(tx: ITxData): Promise<string>;
+}
+
+export function createWalletConnect(): WalletConnect {
+  return new WalletConnectClient();
+}
+
+export class WalletConnectClient implements WalletConnect {
+  private connector: Connector;
+  private disconnectedBus = new Bacon.Bus<void>();
+  // Mutex to synchronize connection and disconnection.
+  private connectionMutex = new Mutex();
+
+  public disconnected: Bacon.EventStream<void>;
+
+  constructor() {
+    this.connector = createConnector();
+    this.disconnected = this.disconnectedBus.toEventStream();
+    this.reinit();
+  }
+
+  getConnection(): Connection | undefined {
+    if (this.connector.connected) {
+      return {
+        chainId: this.connector.chainId,
+        accountAddress: this.connector.accounts[0],
+      };
+    } else {
+      return undefined;
+    }
+  }
+
+  async connect(): Promise<boolean> {
+    return tryRunExclusive(this.connectionMutex, async () => {
+      try {
+        await this.connector.connect();
+        return true;
+      } catch (e) {
+        this.reinit();
+        if (e.message.includes("User close")) {
+          return false;
+        } else {
+          throw e;
+        }
+      }
+    });
+  }
+
+  async disconnect(): Promise<void> {
+    return tryRunExclusive(this.connectionMutex, async () => {
+      await this.connector.killSession().catch(() => {
+        // When the user disconnects wallet-side, calling `killSession`
+        // app-side trows an error because the wallet has already closed
+        // its socket. Therefore, we simply ignore it.
+      });
+      this.reinit();
+    });
+  }
+
+  signMessage(address: string, message: string): Promise<string> {
+    return this.connector.signMessage([address, message]);
+  }
+
+  sendTransaction(tx: ITxData): Promise<string> {
+    return this.connector.sendTransaction(tx);
+  }
+
+  signTransaction(tx: ITxData): Promise<string> {
+    return this.connector.sendTransaction(tx);
+  }
+
+  // We need to reinitialize `WalletConnect` until this issue is fixed:
+  // https://github.com/WalletConnect/walletconnect-monorepo/pull/370
```

Comment:
This PR seems to be merged now, do we still need to re-init now?

Reply:
Yes, I’ve updated the comment. in https://github.com/radicle-dev/radicle-upstream/pull/2123/commits/0e406d57a60c2de4021960f4043e352ddd9eb39a